src/core/image_utils.js

```diff
@@ -0,0 +1,184 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint no-var: error */
+
+import { assert, OPS, warn } from '../shared/util';
+import { ColorSpace } from './colorspace';
+import { JpegStream } from './jpeg_stream';
+import { Stream } from './stream';
+
+class NativeImageDecoder {
+  constructor({ xref, resources, handler, forceDataSchema = false,
+                pdfFunctionFactory, }) {
+    this.xref = xref;
+    this.resources = resources;
+    this.handler = handler;
+    this.forceDataSchema = forceDataSchema;
+    this.pdfFunctionFactory = pdfFunctionFactory;
+  }
+
+  canDecode(image) {
+    return image instanceof JpegStream &&
+           NativeImageDecoder.isDecodable(image, this.xref, this.resources,
+                                          this.pdfFunctionFactory);
+  }
+
+  decode(image) {
+    // For natively supported JPEGs send them to the main thread for decoding.
+    const dict = image.dict;
+    let colorSpace = dict.get('ColorSpace', 'CS');
+    colorSpace = ColorSpace.parse(colorSpace, this.xref, this.resources,
+                                  this.pdfFunctionFactory);
+
+    return this.handler.sendWithPromise('JpegDecode', [
+      image.getIR(this.forceDataSchema), colorSpace.numComps
+    ]).then(function({ data, width, height, }) {
+      return new Stream(data, 0, data.length, dict);
+    });
+  }
+
+  /**
+   * Checks if the image can be decoded and displayed by the browser without any
+   * further processing such as color space conversions.
+   */
+  static isSupported(image, xref, res, pdfFunctionFactory) {
+    const dict = image.dict;
+    if (dict.has('DecodeParms') || dict.has('DP')) {
+      return false;
+    }
+    const cs = ColorSpace.parse(dict.get('ColorSpace', 'CS'), xref, res,
+                                pdfFunctionFactory);
+    // isDefaultDecode() of DeviceGray and DeviceRGB needs no `bpc` argument.
+    return (cs.name === 'DeviceGray' || cs.name === 'DeviceRGB') &&
+           cs.isDefaultDecode(dict.getArray('Decode', 'D'));
+  }
+
+  /**
+   * Checks if the image can be decoded by the browser.
+   */
+  static isDecodable(image, xref, res, pdfFunctionFactory) {
+    const dict = image.dict;
+    if (dict.has('DecodeParms') || dict.has('DP')) {
+      return false;
+    }
+    const cs = ColorSpace.parse(dict.get('ColorSpace', 'CS'), xref, res,
+                                pdfFunctionFactory);
+    const bpc = dict.get('BitsPerComponent', 'BPC') || 1;
+    return (cs.numComps === 1 || cs.numComps === 3) &&
+           cs.isDefaultDecode(dict.getArray('Decode', 'D'), bpc);
+  }
+}
+
+const ImageCacheKind = {
+  NON_EXISTING: 0,
+  EXISTING: 1,
+  NEEDS_RESIZING: 2,
+};
+
+class ImageCache {
```

Comment:
If would be good to have some unit tests for this new class. I think only the handler needs to be mocked for that.

Reply:
Sorry, but adding unit-tests would take time that I really don't have (and quite frankly don't want to spend *now*, since things may change subtly in the follow-up), especially since I still need to do additional work to actually wire-up everything in PR #9255.

Reply:
Also, I purposely adding lots of non-production/test-only `assert`s such that there *should* be a fair amount of implicit test-coverage already through the unit and reference tests.